Ajout de la liste de relecteurs dans les contenus

sylafrs opened the pull request:
<!-- Veuillez décrire vos changements à l’emplacement de ce commentaire (inutile dans le cas de tout petits changements). -->

Ajout de la liste de testeurs dans les contenus : 
- Ajout de la liste dans la vue des tutoriaux
- Ajout d'un bouton "Ajouter un testeur" dans le menu
- Ajout d'un bouton "Gérer les testeurs" dans le menu
- Ajout d'un champs "testers" many to many. (+ migration)
- Ajout des classes AddTesterToContent et RemoveTesterFromContent
- Ajout des formulaires TesterForm et RemoveTesterForm
- Modification du code d'accès à un tutoriel : les testeurs peuvent les lire mais pas les éditer
- Ajout d'un filtre "à tester" dans la liste des tutoriaux d'un utilisateur
- Ajout d'un préfixe [A tester] dans les noms des tutoriaux où nous sommes testeurs
- Ajout d'un filtre django "is_testing" à appliquer sur un contenu, prenant utilisateur, retournant vrai si l'utilisateur est un testeur du contenu
- Ajout d'un filtre django "displayable_testers" reprenant le code de "displayable_authors" pour les testeurs
- Ajout du message privé d'ajout à la liste des testeurs (add_tester_pm.md)


<!-- Indiquez ci-dessous les numéros des tickets (avec le # au début) corrigés par vos changements : -->
Numéro du ticket concerné (optionnel) : #5296 

à noter : ce commit ajoute une liste de testeurs, pas la fonctionnalité du lien temporaire (ou permanent ?)

Attention: Aucun test n'a été rédigé, je risque d'avoir besoin d'aide pour ça
Attention (2) : Pour le moment, il manque une icône pour le filtre "à tester" (une loupe ou un truc comme ça pourrait faire l'affaire) : quel nom d'icone dois-je utiliser ?

sylafrs committed:
Ajout de la liste de testeurs dans les contenus

sylafrs commented:
Je tenais à partager ça avant pour recueillir vos avis (et votre aide pour les tests/le design) : est-ce que ça convient dans l'idée ou est-ce que j'ai tout faux ?

Situphen commented on zds/tutorialv2/views/contents.py:
```diff
@@ -1776,6 +1840,10 @@ def form_valid(self, form):
                 UserGallery(gallery=self.object.gallery, user=user, mode=GALLERY_WRITE).save()
+                if user.pk in all_testers_pk:
+                    self.object.testers.remove(user)
+                    all_testers_pk.remove(user.pk)
```
Je ne pense pas que cette ligne soit utile.

sylafrs replied:
Elle permet de retirer de la liste des testeurs les nouveaux auteurs (pour éviter d'avoir un membre présent dans les deux listes à la fois)

Situphen replied:
C'est le rôle de `self.object.testers.remove(user)`, mais à mon avis `all_testers_pk.remove(user.pk)` ne sert pas à grand chose car c'est une liste qui n'est pas réutilisée après :)

sylafrs replied:
La boucle for itère sur la liste d'utilisateurs que fourni l'utilisateur dans son formulaire.
Si l'utilisateur écrit plusieurs fois le même membre dans la liste, y'a t-il des doublons dans form.cleaned_data['users'] ?

Situphen replied:
Ce n'est pas nécessaire ici car on itère sur la liste des auteurs et il y a déjà un mécanisme pour au cas où il y a deux fois le même pseudo : 

```py
if user.pk not in all_authors_pk and user != self.request.user:
    self.object.authors.add(user)
    all_authors_pk.append(user.pk)
```

mais c'est un détail, ce n'est pas grave si tu laisses cette ligne. :)

sylafrs replied:
Je l'ai retiré au final, si il y a un système qui vérifie en amont, on est bons :)

Situphen commented on zds/tutorialv2/views/contents.py:
```diff
@@ -1787,6 +1855,68 @@ def form_invalid(self, form):
+                    testers_list += _(' et ')
+                else:
+                    testers_list += _(', ')
+            testers_list += user.username
```
On peut faire plus concis, tout en restant compréhensible (à mon avis) : 

```py
testers = form.cleaned_data['users']
testers = ', '.join(testers[:-1]) + ' et ' + testers[-1]

sylafrs replied:
ça marche, on peut aussi appliquer cette modification pour les auteurs du coup

sylafrs replied:
J'ai une question: appeler "user" revient à appeler "user.username" ? Il y a une sorte de ToString() ?

Situphen replied:
Je ne vois pas très bien quelle est ta question. Dans notre code on utilise souvent la variable `user` pour garder un objet User (provenant de la base de données) et `username` pour le pseudo du membre (une simple chaîne de caractères). `username` c'est aussi le nom du champ utilisé dans la base de données pour le pseudo, c'est pourquoi on utilise régulièrement `username = user.username` (que tu peux lire comme `username = <un objet User>.<son champ "username">`). Néanmoins, ce n'est pas toujours respecté et il arrive que le mot "users" soit utilisé à la place de "username", comme pour `form.cleaned_data['users']` qui est une liste de pseudos.

sylafrs replied:
Comment se fait-il que : 
 ```
for index, user in enumerate(form.cleaned_data['users']):
    if index > 0:
        if index == len(users) - 1:
            proofreaders_list += _(' et ')
        else:
            proofreaders_list += _(', ')
    proofreaders_list += user.username
```

fonctionne si `form.cleaned_data['users']` n'est qu'une liste pseudo et pas d'objets utilisateurs ? o.O

sylafrs replied:
La doc précise : "Quelles que soient les données envoyées avec un formulaire, au moment où elles ont été validées avec succès suite à l’appel de is_valid() (et que is_valid() a renvoyé True), les données de formulaire validées se trouvent dans le dictionnaire form.cleaned_data. Ces données auront été gracieusement converties en types Python pour vous."

sylafrs replied:
Oui, du coup je peux confirmer que ce n'est pas une liste de users, j'ai appliqué la modification et reçois une erreur : 

Exception Type: | Exception Value :
-- | --
 TypeError | must be str, not User

Situphen replied:
Je me suis planté ! Je pensais que ça contenait une liste des noms d'utilisateurs et je n'ai pas pris le temps de vérifier en écrivant mon commentaire...

Étant donné que `form.cleaned_data['users']` est bien un liste d'objets `User()`, il faut modifier le code que j'avais donné : 

```py
testers = [u.username for u in form.cleaned_data['users']]
testers = ', '.join(testers[:-1]) + ' et ' + testers[-1]
```

sylafrs replied:
d'accord, je fais ça :)

sylafrs replied:
Presque, voici le message quand on enlève qu'un utilisateur "Vous avez enlevé et user de la liste des relecteurs de ce tutoriel."

Attendu : "Vous avez enlevé user de la liste des relecteurs de ce tutoriel."

sylafrs replied:
```
authors_list = [u.username for u in users]
authors_list = authors_list[0] if len(authors_list) == 1 else ', '.join(authors_list[:-1]) + ' et ' + authors_list[-1]
```

Situphen commented on zds/tutorialv2/views/contents.py:
```diff
@@ -1871,13 +2001,15 @@ class ContentOfAuthor(ZdSPagingListView):
-        ('public', [lambda q: q.filter(sha_public__isnull=False), _('Publiés'), True, 'tick green']),
-        ('validation', [lambda q: q.filter(sha_validation__isnull=False), _('En validation'), False, 'tick']),
-        ('beta', [lambda q: q.filter(sha_beta__isnull=False), _('En bêta'), True, 'beta']),
+        ('public', [lambda q, u: q.filter(sha_public__isnull=False, authors__pk__in=[u]), _('Publiés'), True, 'tick green']),
```
Étant donné que tu n'as qu'un seul utilisateur :)

```suggestion
        ('public', [lambda q, u: q.filter(sha_public__isnull=False, authors__pk=u), _('Publiés'), True, 'tick green']),
```

sylafrs replied:
Ooh, je vois : https://stackoverflow.com/questions/2198635/what-does-pk-in-mean-in-django
Pardon :p

Situphen commented on zds/tutorialv2/views/contents.py:
```diff
@@ -1871,13 +2001,15 @@ class ContentOfAuthor(ZdSPagingListView):
-        ('validation', [lambda q: q.filter(sha_validation__isnull=False), _('En validation'), False, 'tick']),
-        ('beta', [lambda q: q.filter(sha_beta__isnull=False), _('En bêta'), True, 'beta']),
+        ('public', [lambda q, u: q.filter(sha_public__isnull=False, authors__pk__in=[u]), _('Publiés'), True, 'tick green']),
+        ('validation', [lambda q, u: q.filter(sha_validation__isnull=False, authors__pk__in=[u]), _('En validation'), False, 'tick']),
```
```suggestion
        ('validation', [lambda q, u: q.filter(sha_validation__isnull=False, authors__pk=u), _('En validation'), False, 'tick']),
```

Situphen commented on zds/tutorialv2/views/contents.py:
```diff
@@ -1871,13 +2001,15 @@ class ContentOfAuthor(ZdSPagingListView):
-        ('beta', [lambda q: q.filter(sha_beta__isnull=False), _('En bêta'), True, 'beta']),
+        ('public', [lambda q, u: q.filter(sha_public__isnull=False, authors__pk__in=[u]), _('Publiés'), True, 'tick green']),
+        ('validation', [lambda q, u: q.filter(sha_validation__isnull=False, authors__pk__in=[u]), _('En validation'), False, 'tick']),
+        ('beta', [lambda q, u: q.filter(sha_beta__isnull=False, authors__pk__in=[u]), _('En bêta'), True, 'beta']),
```
```suggestion
        ('beta', [lambda q, u: q.filter(sha_beta__isnull=False, authors__pk=u), _('En bêta'), True, 'beta']),
```

AmauryCarrade commented:
Je ne me prononce pas encore sur le code faute de l'avoir lu en détail, mais sur la terminologie, je trouve que le terme de « relecteur » serait peut-être plus approprié.

Et au passage, merci pour le développement, c'est une fonctionnalité à mon sens plus qu'utile !

Situphen commented on zds/tutorialv2/views/contents.py:
```diff
@@ -1871,13 +2001,15 @@ class ContentOfAuthor(ZdSPagingListView):
+        ('beta', [lambda q, u: q.filter(sha_beta__isnull=False, authors__pk__in=[u]), _('En bêta'), True, 'beta']),
         ('redaction', [
-            lambda q: q.filter(sha_validation__isnull=True, sha_public__isnull=True, sha_beta__isnull=True),
+            lambda q, u: q.filter(sha_validation__isnull=True, sha_public__isnull=True, sha_beta__isnull=True, authors__pk__in=[u]),
```
```suggestion
            lambda q, u: q.filter(sha_validation__isnull=True, sha_public__isnull=True, sha_beta__isnull=True, authors__pk=u),
```

Situphen commented on zds/tutorialv2/views/contents.py:
```diff
@@ -1871,13 +2001,15 @@ class ContentOfAuthor(ZdSPagingListView):
-            lambda q: q.filter(sha_validation__isnull=True, sha_public__isnull=True, sha_beta__isnull=True),
+            lambda q, u: q.filter(sha_validation__isnull=True, sha_public__isnull=True, sha_beta__isnull=True, authors__pk__in=[u]),
             _('Brouillons'), False, 'edit']),
+        ('testing', [lambda q, u: q.filter(testers__pk__in=[u]), _('À tester'), True, 'search-submit'])
```
Actuellement ça prend tous les contenus dont le membre est relecteur, je ne suis pas sûr que ce soit idéal. Comment le relecteur pourra-t-il trier entre les contenus déjà publié et les contenus en cours ? :)

```suggestion
        ('testing', [lambda q, u: q.filter(testers__pk=u), _('À tester'), True, 'search-submit'])
```

sylafrs replied:
En effet, peut-être faire une autre liste serait pertinent ?

sylafrs replied:
Au lieu de tout mettre ensemble

Situphen replied:
On pourrait avoir deux boutons radios : "Auteur" et "Relecteur" pour pouvoir faire des combinaisons avec "Publié", "En validation", "En bêta" et "Brouillon"

sylafrs replied:
Une liste de filtres : "Publié", "En validation", "En bêta" et "Brouillon" avec une autre liste de filtres : "Tous", "Auteur" et "Relecteur" du coup ?

Situphen replied:
Ça me parait être une bonne idée, oui !

sylafrs replied:
Je fais ça

Situphen commented on zds/tutorialv2/views/contents.py:
```diff
@@ -1901,13 +2033,14 @@ def dispatch(self, request, *args, **kwargs):
     def get_queryset(self):
         if self.type in list(TYPE_CHOICES_DICT.keys()):
-            queryset = PublishableContent.objects.filter(authors__pk__in=[self.user.pk], type=self.type)
+            queryset = PublishableContent.objects.filter(Q(authors__pk__in=[self.user.pk])|Q(testers__pk__in=[self.user.pk]), type=self.type)
```
```suggestion
            queryset = PublishableContent.objects.filter(Q(authors__pk=self.user.pk)|Q(testers__pk=self.user.pk), type=self.type)
```

Situphen commented on zds/tutorialv2/mixins.py:
```diff
@@ -266,9 +277,9 @@ def get_context_data(self, **kwargs):
         is_same_version = (not self.sha or self.sha == self.object.sha_draft)
-        context['can_add_something'] = is_allowed and is_same_version
+
+        context['can_add_something'] = (self.is_author or self.is_staff) and is_same_version
```
Pourquoi avoir modifié ? :)

sylafrs replied:
Pendant mes tests, j'ai peut-être oublié de revert cette modification là.

Ensuite, vu que is_allowed n'est utilisée que là, et que je trouve le code plus compréhensible de cette manière, je propose de laisser la modification

D'autant plus que maintenant le nom de la variable n'est plus tellement correct : les testeurs sont autorisés à lire le contenu, mais pas à le modifier, du coup il aurait fallu la renommer "is_edit_allowed" ou "is_allowed_to_edit"

Situphen replied:
OK ça me va

A-312 commented:
Il faudrait utiliser un autre nom que testeur, lecteur ?

Situphen commented on templates/tutorialv2/includes/content_item.part.html:
```diff
@@ -32,6 +34,8 @@
     {% if public_content and online %}
         {{ public_content.title }}
+    {% elif content|is_testing:user %}
+        <b>[À tester]</b> {{ content.title }}
```
Si on part du principe qu'on renomme "testeur" en "relecteur", il faudrait avoir quelque chose comme `[En relecture]`.

Mais, je me demande si c'est vraiment important d'avoir ça dans le titre ? Si un membre veut accéder à la liste des contenus dont il est relecteur, il peut aller dans "Mes articles" (ou équivalent) et filter avec "Relecteur".

sylafrs replied:
Je suis d'accord, mais je pense qu'il faudrait carrément une liste à part entière pour que les autres filtres marchent bien

Situphen replied:
Ou sinon on utilise des boutons radios donc il n'y a pas de mélange entre contenus dont on est l'auteur et contenus dont on est le relecteur https://github.com/zestedesavoir/zds-site/pull/5297#discussion_r263279781 :)

sylafrs replied:
-ou- si on utilise une autre liste de filtre, soit retirer soit trouver une autre forme de feedback (un bandeau sur l'image ?)

sylafrs replied:
Du coup je vais retirer

Situphen reviewed (changes requested):
Salut @sylafrs,

Merci d'avoir pris en charge ce ticket ! J'ai fait une relecture de ton code et tu as l'air d'avoir bien compris le fonctionnement du cadriciel (*framework*) Django et de notre code, surtout que c'est ta première PR pour ZDS !

Comme l'a mentionné @AmauryCarrade, je pense que le terme "testeur" (et son équivalent "tester" en anglais) est assez flou. J'aime bien la proposition d'utiliser "relecteur" à la place.

J'ai fait quelques commentaires sur des petits bouts du code, je te laisse les lire et y répondre si besoin. J'ai quelques points plus généraux dont je préfère discuter dans ce message.

- Dans le fichier `zds/tutorialv2/mixins.py`, je me demande si les bouts que code comme ça ne pourraient pas être améliorés (j'ai peur que ce soit difficilement compréhensible pour quelqu'un qui rentre dans le code) : 
  
  ```py
  if ... :
      if ... :
          raise PermissionDenied
  ```
  
  Es-tu du même avis ? Si oui, il reste à trouver comment améliorer ça.

- Dans le fichier `zds/tutorialv2/forms.py`, il y a beaucoup de similitudes entre les formulaires pour les auteurs et ceux pour les relecteurs. On peut prendre par exemple `AuthorForm` et `TesterForm`, à la vue de leur similitude je pense qu'on pourrait simplement avoir : 
  
  ```py
  class TesterForm(AuthorForm):
      username = forms.CharField(
          label=_("Testeurs à ajouter séparés d'une virgule."),
          required=True
      )
  ```
  
  C'est-à-dire partir de `AuthorForm` et ne modifier que le champ `username`. Il en va de même pour `RemoveAuthorForm` et `RemoveTesterForm`

- Dans le fichier ` zds/tutorialv2/views/contents.py`, là encore il y a beaucoup de similitude entre les vues de gestion des auteurs et celles que gestion des relecteurs. C'est peut-être plus compliqué que pour les formulaires mais si on peut réutiliser des bouts de code (pour ne pas réinventer la roue) sans pour autant gêner la compréhension du code, ce serait bien ! :)

Bon, je pense avoir fait le tour ! N'hésite pas si tu as des questions, et encore merci pour ta contribution :)

sylafrs commented:
Par rapport au fait de faire hériter TesterForm d'AuthorForm : je ne suis pas très fan; peut-être passer par une classe commune entre les deux : TesterForm et AuthorForm hérite de TutorialMemberForm par exemple ?

un testeur n'étant pas un auteur

-----------------

Par rapport aux : 

```
if ... :
    if ... :
        raise PermissionDenied
```

Peut-être que créer des booleans en amont pourrait aider ?
Mais personnellement, je suis justement un dev qui vient de rentrer dans le code du ZdS (pire : c'est la première fois que je fais du django et même du python), et j'ai trouvé ça assez clair

-----------------

Par rapport au nommage : "testeur", j'ai utilisé ce terme car on le retrouve dans "Beta-test".
J'ai supposé que l'on pouvais considérer la lecture fermée comme de l'alpha-test d'où le nom testeur

Le nom relecteur va aussi très bien, mais je vois mal comment l'appeler en anglais par contre ^^
proofreader ?

-----------------

Merci beaucoup pour la relecture :)
Je ferais les modifications dès que possible :)

Situphen commented:
> Par rapport au fait de faire hériter TesterForm d'AuthorForm : je ne suis pas très fan; peut-être passer par une classe commune entre les deux : TesterForm et AuthorForm hérite de TutorialMemberForm par exemple ?
> 
> un testeur n'étant pas un auteur

Certes, un relecteur n'est pas un auteur, mais les formulaires pour les ajouter et les enlever sont identiques à l'exception du nom du champ et du nom de la classe ! C'est pour ça que je ne trouve pas choquant de faire hériter `TesterForm` de `AuthorForm`. Je trouve que `TutorialMemberForm` n'est pas forcément très parlant comme nom, mais cette solution peut très bien fonctionner aussi. :)

-----

> Peut-être que créer des booleans en amont pourrait aider ?

Je pense qu'il y a moyen de rendre le code plus compréhensible avec des booléens, mais il faut essayer et voir si c'est réalisable ou pas. :)

-----

> Le nom relecteur va aussi très bien, mais je vois mal comment l'appeler en anglais par contre ^^
proofreader ?

Je pense que *proofreader* correspond bien à ce que l'on souhaite. @AmauryCarrade, un avis sur la question ?

sylafrs commented:
> Certes, un relecteur n'est pas un auteur, mais les formulaires pour les ajouter et les enlever sont identiques à l'exception du nom du champ et du nom de la classe !

Pour le moment: c'est vrai, dans le futur: on ne sait pas

TutorialUserForm  PublishableContentUserForm PublishableContentMemberForm

sylafrs committed:
Correction: Utilisation de pk__in non requise (voir #5297)

sylafrs committed:
Renomme 'Testeur' (tester) en 'Relecteur' (proofreader), voir #5297

sylafrs committed:
Utilise l'héritage pour éviter la duplication de code. (voir #5297)

Nouvelles classes: PublishableContentContributorForm et RemovePublishableContentContributorForm

Situphen commented on zds/tutorialv2/views/contents.py:
```diff
@@ -1732,6 +1736,65 @@ def form_valid(self, form):
+                if user.pk in all_authors_pk:
+                    messages.error(self.request, _('Le membre sélectionné fait partie des auteurs.').format(_type))
+                    return redirect(self.object.get_absolute_url())
+                else:
```
Étant donné qu'il y a un `if ... : return` le `else` n'est pas nécessaire et rajoute une indentation. C'est un détail et je te laisse le choix de l'enlever ou le laisser, mais comme tu as pu voir dans le code c'est assez fréquent de voir : 

```py
def truc():
    if foo : # ce dont on ne veut pas
        return False # ou raise PermissionDenied ou autre

    # longues lignes de code
    return True
```

sylafrs replied:
en effet

sylafrs committed:
Retire un else inutile (voir #5297)

sylafrs committed:
Double filtrage

sylafrs commented on templates/tutorialv2/index.html:
```diff
@@ -172,19 +172,32 @@ <h2 class="ico-after ico-tutorials">
                     {% endif %}
                 {% endfor %}
             </ul>
+            <h4>Par ..</h4>
```
Je n'ai pas trouvé le bon mot.. Une suggestion ?

sylafrs replied:
J'ai enlevé les "Par" et j'ai mis "Status" et "Rôle"

sylafrs committed:
Merge branch 'dev' into feature_tutorial_testers

sylafrs commented:
Que pensez-vous de changer

```
if self.must_be_author and not self.is_author:
    if not self.authorized_for_staff or (self.authorized_for_staff and not self.is_staff):
        raise PermissionDenied

if self.must_be_author_or_proofreader and not self.is_author and not self.is_proofreader:
    if not self.authorized_for_staff or (self.authorized_for_staff and not self.is_staff):
        raise PermissionDenied

```
Par : 
```
        read_authorization = True

        if self.must_be_author:
            read_authorization = read_authorization and self.is_author

        if self.must_be_author_or_proofreader:
            read_authorization = read_authorization and self.is_proofreader
            
        if self.authorized_for_staff:
            read_authorization = self.is_staff
            
        if not read_authorization:
            raise PermissionDenied
```

sylafrs committed:
Fix: string_concat non défini

Le problème semble être apparu depuis le merge.
J'ai récupéré la modification qui avait été faite côté auteurs : utilisation de format_lazy.

sylafrs committed:
Filtres : nom 'Rôle' et icônes

sylafrs committed:
Fix: Migration

sylafrs committed:
Compression du code

sylafrs committed:
Retire une ligne inutile

sylafrs committed:
Retire le prefixe [À relire]

sylafrs commented:
Ensuite il faudra que je corrige les tests existants (renommage de filter) et que je crée les tests pour le code ajouté.

Pour la correction ce sera simple, pour l'ajout, j'avoue avoir besoin d'aide: y'a t-il une bonne ressource en ligne que vous me conseillerez ?

A-312 commented:
Les parties à ne pas valider sont-elles accessible par les relecteurs ?

artragis commented:
@A-312 excellente question, et ma réponse de "maintainer" sera : oui.

sylafrs commented:
Pour moi : oui
Ce n'est pas prêt à être validé, mais ça peut être relu pour suggestion (justement!)

sylafrs commented:
Une question: comment on fait pour retirer le filtrage dans la liste de contenus ?
Si impossible: je propose d'ajouter un lien "Tous"

sylafrs committed:
Ajoute une lien 'Tous' pour réinitialiser les filtres

sylafrs commented:
Ressemble à ça : 

![image](https://user-images.githubusercontent.com/967253/54028727-50f32780-41a6-11e9-93d4-d9051099fdf5.png)

sylafrs committed:
Fix: correction du filtre pour les opinions

sylafrs committed:
Filter modifié dans les tests unitaires

sylafrs commented:
```
Traceback (most recent call last):
  File "/home/travis/build/zestedesavoir/zds-site/zds/tutorialv2/tests/tests_views/tests_content.py", line 668, in test_beta_workflow
    self.assertEqual(result.status_code, 403)  # (get 403 since he is not part of the authors)
AssertionError: 200 != 403
```

En effet le code est buggé : le contenu est disponible en lecture seule pour tout utilisateur connecté ayant l'adresse..
le raise PermissionDenied doit avoir un soucis, je corrige ça

sylafrs commented:
Avec la modification de code proposée ci dessus: ça marche. Je vais la garder du coup.

sylafrs committed:
Fix: Correction du système de droits

sylafrs committed:
Correction: autorisations

sylafrs commented:
à partir de là, je risque d'avoir besoin d'aide :
- Relecture du nouveau code
- Aide par rapport au test qui ne passe pas (j'ai du mal à savoir pourquoi pour le test avec le sha_beta, par exemple)
- Aide pour l'écriture des tests

sylafrs commented:
Il semblerait y avoir un problème avec les versions 'beta'.

https://github.com/zestedesavoir/zds-site/blob/dev/zds/tutorialv2/tests/tests_views/tests_content.py

FAILS:

L725
```
        # get 200 everywhere :)
        result = self.client.get(
            reverse('content:view', args=[tuto.pk, tuto.slug]) + '?version=' + current_sha_beta,
            follow=False)
        self.assertEqual(result.status_code, 200)
```

L3111
```
  # check if user can warn typo in tutorial
        result = self.client.post(
            reverse('content:warn-typo') + '?pk={}'.format(tuto.pk),
            {
                'pk': tuto.pk,
                'version': sha_beta,
                'text': typo_text,
                'target': ''
            },
            follow=True)
        self.assertEqual(result.status_code, 200)
```

sylafrs commented on zds/tutorialv2/mixins.py:
```diff
@@ -114,10 +129,24 @@ def get_versioned_object(self):
+            read_authorization = True
+
+        if not read_authorization:
+            raise PermissionDenied
```
Code soupçonné pour les fails (voir commentaire au dessus)

sylafrs committed:
Correction pour la vue beta.

artragis commented:
> Je suppose que c'est normal, en revanche, qu'un utilisateur non inscrit n'ait pas accès à une beta ?

c'est le but de la béta: seuls les inscrits y ont accès pour faire des retours.

A-312 commented:
> 
> 
> Une question: comment on fait pour retirer le filtrage dans la liste de contenus ?
> Si impossible: je propose d'ajouter un lien "Tous"

#5285

A-312 commented:
> Une question: comment on fait pour retirer le filtrage dans la liste de contenus ?
> Si impossible: je propose d'ajouter un lien "Tous"

Ceci correspond à l'issue : #5285 Je verrais plus le terme "Enlever" ou "désactiver les filtres". Ensuite en icône il y a : `cross` qui semble plus adapté.

sylafrs commented:
> c'est le but de la béta: seuls les inscrits y ont accès pour faire des retours.

Donc ça m'embête, c'est le comportement que je vois quand je teste, mais je n'ai pas ce comportement avec le script de test..

artragis commented:
C'est à dire @sylafrs? avec ta pr la béta est publique? si c'est le cas c'est un bug en effet.

sylafrs commented:
Désolé j'ai été flou :
J'ai le bon comportement dans tous les cas : une personne non inscrite ne peut pas voir le tutoriel.
J'ai cru que le test faisait ça, mais il ne serait pas passé avant je suppose.

Actuellement, quand je vérifie le comportement des droits : tout se passe bien, mais quand le script le teste, il y a un moment où, alors qu'il devrait avoir accès à un tutoriel en beta, il obtient une erreur 403, et je ne sais pas comment ça se fait..

artragis commented:
ok, je regarde demain si je peux faire aller le debugger

artragis commented:
Alors j'ai pu regarder et le verdict : 

- on a en effet nos tests sur la partie "béta" qui ne sont pas à jour
- tu avais un petit bug tout de même sur la partie "visiter une partie d'un tuto en béta".

J'ai un diff pour corriger tout ça (et faire fonctionner le test sur windows, en plus :p) 

```diff
diff --git a/zds/tutorialv2/mixins.py b/zds/tutorialv2/mixins.py
index 3293212fd..f3b9f8e2b 100644
--- a/zds/tutorialv2/mixins.py
+++ b/zds/tutorialv2/mixins.py
@@ -91,7 +91,6 @@ class SingleContentViewMixin(object):
         self.is_author = self.request.user in obj.authors.all()
         self.is_proofreader = self.request.user in obj.proofreaders.all()
 
-
         read_authorization = True
 
         if self.must_be_author:
diff --git a/zds/tutorialv2/models/database.py b/zds/tutorialv2/models/database.py
index ce04444c7..97d679dcd 100644
--- a/zds/tutorialv2/models/database.py
+++ b/zds/tutorialv2/models/database.py
@@ -51,7 +51,8 @@ class PublishableContent(models.Model, TemplatableContentModelMixin):
 
     A PublishableContent retains metadata about a content in database, such as
 
-    - authors, proofreaders, description, source (if the content comes from another website), subcategory, tags and licence ;
+    - authors, proofreaders, description, source (if the content comes from another website), subcategory,
+    tags and licence ;
     - Thumbnail and gallery ;
     - Creation, publication and update date ;
     - Public, beta, validation and draft sha, for versioning ;
@@ -390,8 +391,8 @@ class PublishableContent(models.Model, TemplatableContentModelMixin):
         """
 
         attrs = [
-            'pk', 'authors', 'proofreaders', 'subcategory', 'image', 'creation_date', 'pubdate', 'update_date', 'source',
-            'sha_draft', 'sha_beta', 'sha_validation', 'sha_public', 'tags', 'sha_picked', 'converted_to',
+            'pk', 'authors', 'proofreaders', 'subcategory', 'image', 'creation_date', 'pubdate', 'update_date',
+            'source', 'sha_draft', 'sha_beta', 'sha_validation', 'sha_public', 'tags', 'sha_picked', 'converted_to',
             'type'
         ]
 
diff --git a/zds/tutorialv2/tests/tests_lists.py b/zds/tutorialv2/tests/tests_lists.py
index 2f4b30b0b..804c55d57 100644
--- a/zds/tutorialv2/tests/tests_lists.py
+++ b/zds/tutorialv2/tests/tests_lists.py
@@ -69,14 +69,17 @@ class ContentTests(TutorialTestMixin, TestCase):
         resp = self.client.get(reverse('content:find-tutorial', args=[self.user_author.pk]) + '?filter_status=public')
         self.assertContains(resp, tutorial.title)
         self.assertNotContains(resp, tutorial_unpublished.title)
-        resp = self.client.get(reverse('content:find-tutorial', args=[self.user_author.pk]) + '?filter_status=redaction')
+        resp = self.client.get(reverse('content:find-tutorial',
+                                       args=[self.user_author.pk]) + '?filter_status=redaction')
         self.assertEqual(resp.status_code, 403)
-        resp = self.client.get(reverse('content:find-article', args=[self.user_author.pk]) + '?filter_status=public')
+        resp = self.client.get(reverse('content:find-article',
+                                       args=[self.user_author.pk]) + '?filter_status=public')
         self.assertContains(resp, article.title)
         self.assertNotContains(resp, article_unpublished.title)
         resp = self.client.get(reverse('content:find-article', args=[self.user_author.pk]) + '?filter_status=redaction')
         self.assertEqual(resp.status_code, 403)
-        resp = self.client.get(reverse('content:find-article', args=[self.user_author.pk]) + '?filter_status=chuck-norris')
+        resp = self.client.get(reverse('content:find-article',
+                                       args=[self.user_author.pk]) + '?filter_status=chuck-norris')
         self.assertEqual(resp.status_code, 404)
 
     def _create_and_publish_type_in_subcategory(self, content_type, subcategory):
diff --git a/zds/tutorialv2/tests/tests_views/tests_content.py b/zds/tutorialv2/tests/tests_views/tests_content.py
index 30e9b370f..fc5941bd0 100644
--- a/zds/tutorialv2/tests/tests_views/tests_content.py
+++ b/zds/tutorialv2/tests/tests_views/tests_content.py
@@ -724,28 +724,28 @@ class ContentTests(TutorialTestMixin, TestCase):
 
         # get 200 everywhere :)
         result = self.client.get(
-            reverse('content:view', args=[tuto.pk, tuto.slug]) + '?version=' + current_sha_beta,
+            reverse('content:beta-view', args=[tuto.pk, tuto.slug]),
             follow=False)
         self.assertEqual(result.status_code, 200)
 
         result = self.client.get(
-            reverse('content:view-container',
+            reverse('content:beta-view-container',
                     kwargs={
                         'pk': tuto.pk,
                         'slug': tuto.slug,
                         'container_slug': self.part1.slug
-                    }) + '?version=' + current_sha_beta,
+                    }),
             follow=False)
         self.assertEqual(result.status_code, 200)
 
         result = self.client.get(
-            reverse('content:view-container',
+            reverse('content:beta-view-container',
                     kwargs={
                         'pk': tuto.pk,
                         'slug': tuto.slug,
                         'parent_container_slug': self.part1.slug,
                         'container_slug': self.chapter1.slug
-                    }) + '?version=' + current_sha_beta,
+                    }),
             follow=False)
         self.assertEqual(result.status_code, 200)
 
@@ -818,7 +818,7 @@ class ContentTests(TutorialTestMixin, TestCase):
         self.assertEqual(result.status_code, 403)  # no access using the old version
 
         result = self.client.get(
-            reverse('content:view', args=[tuto.pk, tuto.slug]) + '?version=' + current_sha_beta,
+            reverse('content:beta-view', args=[tuto.pk, tuto.slug]),
             follow=False)
         self.assertEqual(result.status_code, 200)  # ok for the new version
 
@@ -890,7 +890,7 @@ class ContentTests(TutorialTestMixin, TestCase):
         self.assertEqual(result.status_code, 403)  # no access on the non-beta version (of course)
 
         result = self.client.get(
-            reverse('content:view', args=[tuto.pk, tuto.slug]) + '?version=' + old_sha_beta,
+            reverse('content:beta-view', args=[tuto.pk, tuto.slug]),
             follow=False)
         self.assertEqual(result.status_code, 200)  # access granted
 
diff --git a/zds/tutorialv2/urls/urls_contents.py b/zds/tutorialv2/urls/urls_contents.py
index 310a55ca6..27e04a831 100644
--- a/zds/tutorialv2/urls/urls_contents.py
+++ b/zds/tutorialv2/urls/urls_contents.py
@@ -140,10 +140,10 @@ urlpatterns = [
     re_path(r'^ajouter-auteur/(?P<pk>\d+)/$',
             AddAuthorToContent.as_view(), name='add-author'),
     re_path(r'^enlever-auteur/(?P<pk>\d+)/$',
-            RemoveAuthorFromContent.as_view(), name='remove-author'),  
-    re_path(r'^ajouter-relecteur/(?P<pk>\d+)/$', 
-            AddProofreaderToContent.as_view(), name='add-proofreader'),  
-    re_path(r'^enlever-relecteur/(?P<pk>\d+)/$', 
+            RemoveAuthorFromContent.as_view(), name='remove-author'),
+    re_path(r'^ajouter-relecteur/(?P<pk>\d+)/$',
+            AddProofreaderToContent.as_view(), name='add-proofreader'),
+    re_path(r'^enlever-relecteur/(?P<pk>\d+)/$',
             RemoveProofreaderFromContent.as_view(), name='remove-proofreader'),
     # beta:
     re_path(r'^activer-beta/(?P<pk>\d+)/(?P<slug>.+)/$', ManageBetaContent.as_view(action='set'),
diff --git a/zds/tutorialv2/utils.py b/zds/tutorialv2/utils.py
index ba6adf832..5d140aa7b 100644
--- a/zds/tutorialv2/utils.py
+++ b/zds/tutorialv2/utils.py
@@ -43,9 +43,9 @@ def search_container_or_404(base_content, kwargs_array):
 
     if isinstance(kwargs_array, str):
         dic = {}
-        dic['parent_container_slug'] = kwargs_array.split('/')[0]
-        if len(kwargs_array.split('/')) >= 2:
-            dic['container_slug'] = kwargs_array.split('/')[1]
+        dic['parent_container_slug'] = kwargs_array.split(os.path.sep)[0]
+        if len(kwargs_array.split(os.path.sep)) >= 2:
+            dic['container_slug'] = kwargs_array.split(os.path.sep)[1]
         kwargs_array = dic
 
     if 'parent_container_slug' in kwargs_array:
diff --git a/zds/tutorialv2/views/contents.py b/zds/tutorialv2/views/contents.py
index 4d0b79fc5..8a64d63e1 100644
--- a/zds/tutorialv2/views/contents.py
+++ b/zds/tutorialv2/views/contents.py
@@ -36,8 +36,8 @@ from zds.member.models import Profile
 from zds.notification.models import TopicAnswerSubscription, NewPublicationSubscription
 from zds.tutorialv2.forms import ContentForm, JsFiddleActivationForm, AskValidationForm, AcceptValidationForm, \
     RejectValidationForm, RevokeValidationForm, WarnTypoForm, ImportContentForm, ImportNewContentForm, ContainerForm, \
-    ExtractForm, BetaForm, MoveElementForm, AuthorForm, RemoveAuthorForm, ProofreaderForm, RemoveProofreaderForm, CancelValidationForm, PublicationForm, \
-    UnpublicationForm
+    ExtractForm, BetaForm, MoveElementForm, AuthorForm, RemoveAuthorForm, ProofreaderForm, RemoveProofreaderForm,\
+    CancelValidationForm, PublicationForm, UnpublicationForm
 from zds.tutorialv2.mixins import SingleContentDetailViewMixin, SingleContentFormViewMixin, SingleContentViewMixin, \
     SingleContentDownloadViewMixin, SingleContentPostMixin, FormWithPreview
 from zds.tutorialv2.models import TYPE_CHOICES_DICT
@@ -1044,6 +1044,7 @@ class DisplayBetaContainer(DisplayContainer):
     """View to get the beta version of a container"""
 
     sha = None
+    must_be_author_or_proofreader = False
 
     def get_object(self, queryset=None):
         """rewritten to ensure that the version is set to beta, raise Http404 if there is no such version"""
@@ -1899,7 +1900,10 @@ class RemoveProofreaderFromContent(LoggedWithReadWriteHability, SingleContentFor
         self.object.save(force_slug_update=False)
 
         proofreaders_list = [u.username for u in users]
-        proofreaders_list = proofreaders_list[0] if len(proofreaders_list) == 1 else ', '.join(proofreaders_list[:-1]) + ' et ' + proofreaders_list[-1]
+        if len(proofreaders_list) == 1:
+            proofreaders_list = proofreaders_list[0]
+        else:
+            proofreaders_list = ', '.join(proofreaders_list[:-1]) + ' et ' + proofreaders_list[-1]
 
         messages.success(
             self.request, _('Vous avez enlevé {} de la liste des relecteurs de {}.').format(proofreaders_list, _type))
@@ -1912,6 +1916,7 @@ class RemoveProofreaderFromContent(LoggedWithReadWriteHability, SingleContentFor
         self.success_url = self.object.get_absolute_url()
         return super(RemoveProofreaderFromContent, self).form_valid(form)
 
+
 class RemoveAuthorFromContent(LoggedWithReadWriteHability, SingleContentFormViewMixin):
 
     form_class = RemoveAuthorForm
@@ -1965,7 +1970,10 @@ class RemoveAuthorFromContent(LoggedWithReadWriteHability, SingleContentFormView
         self.object.save(force_slug_update=False)
 
         authors_list = [u.username for u in users]
-        authors_list = authors_list[0] if len(authors_list) == 1 else ', '.join(authors_list[:-1]) + ' et ' + authors_list[-1]
+        if len(authors_list) == 1:
+            authors_list = authors_list[0]
+        else:
+            authors_list =', '.join(authors_list[:-1]) + ' et ' + authors_list[-1]
 
         if not current_user:  # if the removed author is not current user
             messages.success(
@@ -2039,7 +2047,8 @@ class ContentOfAuthor(ZdSPagingListView):
 
     def get_queryset(self):
         if self.type in list(TYPE_CHOICES_DICT.keys()):
-            queryset = PublishableContent.objects.filter(Q(authors__pk=self.user.pk)|Q(proofreaders__pk=self.user.pk), type=self.type)
+            queryset = PublishableContent.objects.filter(Q(authors__pk=self.user.pk) | Q(proofreaders__pk=self.user.pk),
+                                                         type=self.type)
         else:
             raise Http404('Ce type de contenu est inconnu dans le système.')
 
@@ -2099,12 +2108,14 @@ class ContentOfAuthor(ZdSPagingListView):
             authorized_filter_status = self.authorized_filters_status[filter_status_]
             if self.user != self.request.user and not authorized_filter_status[2]:
                 continue
-            context['filters_status'].append({'key': filter_status_, 'text': authorized_filter_status[1], 'icon': authorized_filter_status[3]})
+            context['filters_status'].append({'key': filter_status_, 'text': authorized_filter_status[1],
+                                              'icon': authorized_filter_status[3]})
 
         for filter_rights_ in list(self.authorized_filters_rights.keys()):
             authorized_filter_rights = self.authorized_filters_rights[filter_rights_]
             if self.user != self.request.user and not authorized_filter_rights[2]:
                 continue
-            context['filters_rights'].append({'key': filter_rights_, 'text': authorized_filter_rights[1], 'icon': authorized_filter_rights[3]})
+            context['filters_rights'].append({'key': filter_rights_, 'text': authorized_filter_rights[1],
+                                              'icon': authorized_filter_rights[3]})
 
         return context
```

sylafrs commented:
Hmm. Du coup, j'ai lu le diff, mais je ne trouve pas la correction du bug dont tu parles.. 
J'ai vu la modif pour les tests et pour le search_container_or_404

Merci en tout cas : je teste ça ce soir :)

artragis commented:
Elle se trouve dans `DisplayBetaContainer`, tu avais oublié de mettre le `must_be_author_or_proofreader ` à false

sylafrs commented:
Oups, merci !
Je n'ai pas eu le temps de regarder hier soir (pardon), je regarde dès que je peux

sylafrs committed:
Apply diff from artagis (see #5297)

sylafrs commented:
J'ai un bug, je vais regarder ça. 
ïtrema a une tribune : "ullam-tenetur-eaque"

Si j'y accède via la page "tous les billets", je n'ai aucun problème ("billets" dans l'url) -> ListOpinions, DisplayOnlineOpinion
Si j'y accède via la page "tribunes de ïtrema", le contenu m'est refusé avec une erreur 403 ("contenus" dans l'url) -> ContentOfAuthor, DisplayContent

En outre, il faut que je retire le filtrage Auteur/Relecteur à cet endroit.

Le problème doit venir de : SingleContentDetailViewMixin ou bien du fait que l'on devrait sans doute changer les liens de ContentOfAuthor ?

Peut-être créer une classe : "OnlineContentOfAuthor" ?

sylafrs committed:
Merge branch 'dev' into feature_tutorial_testers

artragis commented:
On en est où ici?

sylafrs commented:
J'ai eu pas mal de trucs à faire ces derniers temps (pas eu beaucoup de temps pour moi ^^'), donc j'allais reprendre où j'en étais.
Mon dernier commentaire indique, du coup, l'état du code.

artragis commented:
as-tu besoin d'aide pour continuer ici?

artragis commented:
As-tu besoin d'aide pour continuer ici?

EtoileFilante commented:
Hello ! Du neuf ?

firm1 commented:
Aie, je suis passé a coté de cette PR qui est dans le même esprit que ce que j'ai fait [ici](https://github.com/zestedesavoir/zds-site/pull/5485)

Du coup je suis bien embêté, on fait quoi du coup ?

Situphen commented:
@firm1 Cette PR et #5485 ne sont pas tout à fait équivalente car celle-ci donne un droit de lecture aux relecteurs.

En tous cas, il n'y a pas eu d'activité de @sylafrs depuis fin avril donc je ferme cette PR. Et puis il va y avoir des gros conflits entre vos deux PRs si on garde les deux.

A-312 commented:
Je pensais que Firm1 avait mis des droits aussi

sylafrs commented:
Oui pardon, puis la branche n'a pas été mise à jour, le merge risque d'être compliqué, pardon pour ça ^^"

Situphen commented:
@sylafrs Pas de soucis ! :)
